### DISCUSSION TITLE
Varibale derivates, coupled variable derivatives

### DISCUSSION THREAD

**Discussion by Alirezarei** [idaholab/moose#24580] [Q&A General] [answered]
I have general questions about how to add variable derivatives to the kernel. My specific questions are:

1. How can I add different variable derivatives (let it be WRT time, space, or a mix of these two) to the kernel. Is there any source or sample that I can refer to?!
 Let's say I want to use _ad_grad_u_dot in my kernel. How should the header and the source files look like?

2. How can I do the same for coupled variables and their derivatives?

Thank you,
Alireza

**Comment by GiudGiud** [comment]
Hello

You have to leverage the coupleable (interfaces/coupleable.h) interface to retrieve those derivatives (time -> coupledDot, space -> coupledGradient)
In your header for the kernel you are creating, add references to those values. The type of the values is the return type of the routine in the coupleable interface.

Guillaume

**Reply by Alirezarei** [reply]
Hi Guillaume,

Thank you so much for your response. I'm still struggling with this. Could you please provide some sample header and source files in which they use higher order dedrivatives like ad_grad_u_dot or the second order derivative WRT time and space of the field variables?

It would be great if you can refer me to some sample header and source files that do the same for coupled variables as well.

I appreciate your help,
Alireza

**Reply by GiudGiud** [reply, marked as answer]
Here's one with a u_dot
https://github.com/idaholab/moose/blob/968eb46a3868f624d7930083ffdedd039b3d4fd5/framework/include/kernels/TimeKernel.h

Here s on with a grad_u
https://github.com/idaholab/moose/blob/968eb46a3868f624d7930083ffdedd039b3d4fd5/framework/include/auxkernels/DiffusionFluxAux.h

Second time derivative of a VECTOR (remove vector to get a scalar)
https://github.com/idaholab/moose/blob/968eb46a3868f624d7930083ffdedd039b3d4fd5/modules/electromagnetics/include/kernels/VectorSecondTimeDerivative.h#L30

in general you want to rely on this page to use the coupleable interface:
https://mooseframework.inl.gov/docs/doxygen/moose/classCoupleable.html
